documentation/design-docs/dotnet-tools.md

```diff
@@ -1,124 +1,801 @@
+**Seeing performance counter values that refresh periodically in-place**
 
+    > dotnet tool install -g dotnet-counters
+    You can invoke the tool using the following command: dotnet-counters
+    Tool 'dotnet-counters' (version '1.0.0') was successfully installed.
+    > dotnet counters monitor --process-id 1902 Microsoft-Windows-DotNETRuntime Microsoft-AspNet
+    
+    Microsoft-Windows-DotNETRuntime: 
+        Total Processor Time (ms)              173923.48 
+        Private Virtual Memory (MB)                 1094       
+        Working Set (MB)                            1982
+        Virtual Memory (MB)                         3041
+        GC Heap Memory (MB)                          784
+        Exception Thrown Rate (exceptions/min)       117
+        Lock Contention Rate (contentions/min)      1792               
+
+     Microsoft-AspNet:
+        Request Rate (requests/sec)                 1915
+        Request Latency (ms)                          34  
+       
+    'p' - pause updates
+    'r' - resume updates
+    'q' - quit
+
+(Counter groups are for example purpose only, exact groups/counters TBD)
+
+### Capture a trace for performance analysis
+
+For analyzing CPU usage, IO, lock contention, allocation rate, etc the investigator wants to capture a performance trace. This trace can then be moved to a developer machine where it can be analyzed with profiling tools such as PerfView/VisualStudio or visualized as a flame graph with speedscope. 
+
+**Capture default trace**
+
+    > dotnet tool install -g dotnet-trace
+    You can invoke the tool using the following command: dotnet-trace
+    Tool 'dotnet-trace' (version '1.0.0') was successfully installed.
+    > dotnet trace collect --process-id 1902
+    Recording trace 38MB  
 
+    's' - stop tracing
+    'g' - capture GC heap snapshot
 
+...Hit 's'...
 
-@shirhatti wrote:
+    > dotnet trace collect --process-id 1902
+    Recording trace 107MB
+    Recording complete
+    Trace complete: ~/trace.netperf
 
-```
-NAME
+**Convert a trace to use with speedscope**
+
+    > dotnet trace convert --format speedscope ~/trace.netperf
+    Writing:     ~/trace.speedscope.json
+    Conversion complete
+
+### Do a (dump-based) memory leak analysis
+
+For analyzing managed memory leaks over time, the investigator first wants to capture a series of dumps that will show the memory growth.
+
+    > dotnet tool install -g dotnet-dump
+    You can invoke the tool using the following command: dotnet-dump
+    Tool 'dotnet-dump' (version '1.0.0') was successfully installed.
+    > dotnet dump --process-id 1902 --number 2
+    Writing:     ./core0001
+    
+... 10 seconds pass (the default time interval)
+
+    > dotnet dump --process-id 1902 --number 2
+    Writing:     ./core0001
+    Writing:     ./core0002
+    Complete
+
+Next the investigator needs to compare the heaps in these two dumps.
+
+    > dotnet dump analyze ./core0002
```

Comment:
I think a tool should be a repl or not a repl. Having one command be a repl and others not seems odd.

Reply:
Does it seem better to you if we added a collect command you could use inside the REPL, and then declared the top-level subcommand as a shortcut for launching the REPL and issuing collect from inside? Many command-line debuggers have shortcuts that let a single shell command act the same as if you launched the tool to enter the REPL, entered a command into the REPL, and then exited the REPL.

Another alternative is we break this into two tools, but the value of being able to mentally classify each tool as 'REPL' or 'not-REPL' didn't seem greater than the hassle of being required to remember, install, and use two separate tools for the same task. I don't claim that my opinion on it will wind up reflecting the opinions of our customers though, its just my own educated guess.

I'm open to other suggestions as well if you have them? (Thanks for raising it btw, it never crossed my mind)

Reply:
@richlander Is it really that confusing? I thought of a sub-command launching a REPL to be no different then say when something like `git log` drops you into an interactive pager. Is there any prior art of why we shouldn't combine them.

Reply:
Thinking on it a little more, I'm not liking my suggestion above to have a command from inside the REPL that collects dumps so much. It requires the REPL to operate a state where you may not have opened a dump yet and then we have to define what every other command is going to do in that state. It seems like complexity we can avoid and capturing dumps from inside the REPL doesn't seem to add value.

 I guess I'm basically with @shirhatti, although not that typical, none of the other options I've come up with yet seemed better. I'm open to suggestions or persuasion, but absent hearing something more I don't have a better option to change this to.